### PR TITLE
MINOR; Log reason for deletion of KRaft snapshot

### PR DESCRIPTION
**Pull request by hni61223**
Add better logging for KRaft snapshot deletion. Specifically, add logs to indicate the reason why a snapshot is deleted.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 1]
```diff
@@ -179,7 +180,7 @@ final class KafkaMetadataLog private (
         (false, mutable.TreeMap.empty[OffsetAndEpoch, Option[FileRawSnapshotReader]])
     }
 
-    removeSnapshots(forgottenSnapshots)
+    removeSnapshots(forgottenSnapshots,RetentionMsBreach())
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala]: Missing space between `,` and `RetentionMsBreach`.

File: core/src/main/scala/kafka/MetadataLogConfig.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      config.getLong(KafkaConfig.MetadataLogSegmentMillisProp),
+      config.getLong(KafkaConfig.MetadataMaxRetentionBytesProp),
+      config.getLong(KafkaConfig.MetadataMaxRetentionMillisProp),
+      maxBatchSizeInBytes,
+      maxFetchSizeInBytes,
+      LogConfig.DEFAULT_FILE_DELETE_DELAY_MS,
+      config.getInt(KafkaConfig.NodeIdProp)
+    )
+  }
+}
```
Comment (jsancio) [inline core/src/main/scala/kafka/MetadataLogConfig.scala]: Missing new line.

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 3]
```diff
@@ -348,9 +349,9 @@ final class KafkaMetadataLog private (
           snapshotId.offset <= latestSnapshotId.offset &&
           log.maybeIncrementLogStartOffset(snapshotId.offset, LogStartOffsetIncrementReason.SnapshotGenerated) =>
             // Delete all segments that have a "last offset" less than the log start offset
-            log.deleteOldSegments()
-            // Remove older snapshots from the snapshots cache
-            (true, forgetSnapshotsBefore(snapshotId))
+          log.deleteOldSegments()
+          // Remove older snapshots from the snapshots cache
+          (true, forgetSnapshotsBefore(snapshotId))
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala]: I think this indentation is off by two spaces. I think you want to revert this change.

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 4]
```diff
@@ -359,6 +360,27 @@ final class KafkaMetadataLog private (
...
+          // Remove older snapshots from the snapshots cache
+          val forgottenSnapshots = forgetSnapshotsBefore(snapshotId)
+          (deletedSegments != 0 || forgottenSnapshots.nonEmpty, forgottenSnapshots)
+        case _ =>
+          (false, mutable.TreeMap.empty[OffsetAndEpoch, Option[FileRawSnapshotReader]])
+      }
+    }
+    removeSnapshots(forgottenSnapshots, reason)
+    deleted
+  }
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala]: This looks like a duplicate of `deleteBeforeSnapshot(OffsetAndEpoch)`. If tests are calling this method directly, lets add an "unknown" reason and the test can use that when deleting snapshots.

You should be able to avoid this code duplication with this change:
```scala
override def deleteBeforeSnapshot(snapshotId: OffsetAndEpoch): Boolean = {
  deleteBeforeSnapshot(snapshotId, UnknownReason())
}
```

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 5]
```diff
@@ -484,9 +506,10 @@ final class KafkaMetadataLog private (
    * Rename the given snapshots on the log directory. Asynchronously, close and delete the
    * given snapshots after some delay.
    */
+
   private def removeSnapshots(
-    expiredSnapshots: mutable.TreeMap[OffsetAndEpoch, Option[FileRawSnapshotReader]]
-  ): Unit = {
+     expiredSnapshots: mutable.TreeMap[OffsetAndEpoch, Option[FileRawSnapshotReader]],
+ ): Unit = {
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala]: This indentation doesn't look correct. I think you want to revert this change.

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 6]
```diff
@@ -501,6 +524,24 @@ final class KafkaMetadataLog private (
...
+    }
+
+    if (expiredSnapshots.nonEmpty) {
+      scheduler.scheduleOnce(
+        "delete-snapshot-files",
+        () => KafkaMetadataLog.deleteSnapshotFiles(log.dir.toPath, expiredSnapshots, this),
+        config.fileDeleteDelayMs
+      )
+    }
+  }
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala]: This looks like a duplicate of the method `removeSnapshots(mutable.TreeMap[_, _])`.

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 7]
```diff
@@ -677,4 +692,20 @@ object KafkaMetadataLog extends Logging {
...
     }
   }
+
+  trait SnapshotDeletionReason {
+    def logReason(snapshotId: OffsetAndEpoch): Unit
+  }
+
+  final case class RetentionMsBreach() extends SnapshotDeletionReason {
+    override def logReason(snapshotId: OffsetAndEpoch): Unit = {
+      info(s"Marking snapshot $snapshotId for deletion because the age is too old")
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:668]: Should this function returns a `String` instead log calling `info` directly?

The reason should include this information: `now - timestamp > config.retentionMillis`. From https://github.com/apache/kafka/pull/15450/files#diff-b332f85b04775c821226b6f704e91d51f9647f29ba73dace65b99cf36f6b9ceaR455

For Example,
`s"the snapshot's timestamp ($timestamp) is now ($now) older than the retention (${config.retentionMillis})"`

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 8]
```diff
@@ -677,4 +692,20 @@ object KafkaMetadataLog extends Logging {
...
+    }
+  }
+
+  final case class RetentionSizeBreach() extends SnapshotDeletionReason {
+    override def logReason(snapshotId: OffsetAndEpoch): Unit = {
+      info(s"Marking snapshot $snapshotId for deletion because the size is too big")
+    }
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:674]: Similar comment here. The reason should include this information: `log.size + snapshotTotalSize > config.retentionMaxBytes`

From: https://github.com/apache/kafka/pull/15450/files#diff-b332f85b04775c821226b6f704e91d51f9647f29ba73dace65b99cf36f6b9ceaR477

**Review by jsancio — COMMENTED** [review]
Thanks for the change @hni61223 . Can you please add a description to the PR?

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala (hunk 7 above)
Comment (hni61223) [reply core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:668]: Do you mind giving me a hint on how to pass the time difference (i.e., now - timestamp) to the `RetentionMsBreanch` function? I tried to modify the `shouldClean` function to pass a tuple of (Boolean, Long) to pass the time difference as the second factor, but was blocked. Could you give me some hints to unblock? Thanks!

**Comment by jsancio** [comment]
We merged https://github.com/apache/kafka/pull/15478 instead.
